articles/150_roslaunch.md

```diff
@@ -428,62 +428,179 @@ The remapping design document[^static_remapping] goes into detail on how you can
-  <node package="my_package" executable="listener" />
+This API could have very low latency to launch nodes since it does not require waiting for discovery.
+However, there is no way to get feedback about the success or failure of loaded nodes.
+There is also no way to tell a container process to unload a composable node.
+
+###### API using STDIN
+
+Another option for a container process API is to pass configuration in via STDIN.
+
+Advantages
+* No waiting for an API to become available
+* Works with read-only file systems
+
+Disadvantages
+* Requires parsing a config
+* Cannot tell from the outside if a container process supports this interface
+* Cannot tell if and when nodes are loaded or unloaded
+* Cannot stop dynamically loaded nodes from reading STDIN
+
+This API could also have very low latency to launch nodes.
+However, there also is no way to get feedback about the success or failure of loaded nodes.
+STDOUT cannot be used because a composable node logging messages to STDOUT is assumed to be very common and would conflict.
+Since STDIN is always available, it would be possible to unload a node via this API.
+
+###### API using ROS Services or Topics
+
+Lastly, a container process API may be defined by ROS services or topics.
+
+Advantages
+* No config file parsing
+* Works with read-only file systems
+* Can indicate if a node was successfully loaded
+* Can create API to trigger launch events
+* Can tell if a container process supports this interface
+
+Disadvantages
+* Must wait for the service API to become available
+* Cannot stop dynamically loaded nodes from creating the same services
+
+This is the only option discussed which can communicate the success or failure of dynamically launched nodes.
+It is also the only option that allows introspection.
+However, this option has the highest potential delay from when the container process is spawned to when nodes may be loaded.
+
+##### Proposed Container process API
+
+This is a proposal for a container process API to be used by the launch system.
+
+###### Command Line Arguments
+The container process must accept command line arguments including log level, remapping, and parameters.
+These command line arguments must not be applied to dynamically launched nodes.
+The launch system will pass these arguments to the container process in the same way it would pass them to a node.
+If a remap rule would apply to a launch service, the launch system should try to use the remapped service name instead.
+
+###### ROS Services
+The container process must offer all of the following services.
 
-</node_container_process>
+* `_launch/load_node`
+* `_launch/unload_node`
+* `_launch/list_nodes`
+
+The services are hidden to avoid colliding with user created services.
+`load_node` will be called by the launch system when a composable node is to be dynamically loaded, and `unload_node` destroys a composable node.
+`list_nodes` is not called by launch system, and is only provided for introspection.
```

Comment:
Meta: why is this API launch specific? It looks general enough to be that of any composable nodes container.

Reply:
I never thought of it as a launch-specific API, but I think that's the context that we are currently designing it for, so there's probably a bias towards launch here.

I don't see anything here that would stop people from writing additional node containers that conform to this API, or additional tooling that could spawn nodes in composable containers.

Reply:
I think someone will write a custom container process (i.e. for a custom client library or executor). I made the wording about container processes a little more generic in aa18ecaffc9e7685a22ab9e172c5fbcdbc59d8a0.

Reply:
Changed `_launch` to `~/_containter` in d98bb98b67d3bd5df9af1bb0ec5f2a741e20b80c